[PVR] EpgContainer: use shared_ptr to manage CEpg pointers

MaxKellermann opened the pull request:
This fixes a crash that can occur when one thread is inside
PersistAll() and the other is inside DeleteEpg().  The former does not
lock the mutex, as it operates on a std::map copy with CEpg pointers.
These CEpg instances however are deleted by the other thread.

Background information:

CEpgContainer::m_critSection protects the std::map m_epgs which
contains CEpg pointers.  All access to m_epgs must be protected with
m_critSection.  What the authors of that code didn't consider: if a
CEpg pointer escapes that protected code section, it can be
invalidated at any time.  Which implies: it is illegal to leak CEpg
pointers out of CEpgContainer::m_critSection protection!

That, however, is what commit 2ba1c43 implements, and what has been
done wrong many times before, for example GetById() and GetByChannel()
leak those pointers, explicitly.  All of that is wrong, and each one
of them can crash Kodi.

So what this commit does is ensure that all CEpg instances remain
valid as long as somebody is still holding a pointer, by wrapping all
CEpg pointers in std::shared_ptr.

See https://github.com/xbmc/xbmc/pull/8238 for my previous attempt to get this merged; I was asked to re-request the pull in the master branch instead of fixing this crash bug in the stable branch.

MartijnKaijser commented:
thx :)

fritsch commented:
Jenkins build this please @jalle please take care :-)

The much
Am 18.10.2015 3:17 nachm. schrieb "Martijn Kaijser" <
notifications@github.com>:

> thx :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8247#issuecomment-149006511>.
>

hudokkow commented:
jenkins build this please

fritsch commented:
thx @hudokkow

ksooo commented:
Just a question: If I'm not mistaken, there are still some naked Epg* around after this PR. Is this by intention and if yes, why? Thx.

MaxKellermann commented:
> there are still some naked Epg* around after this PR
Non-owning pointers are still possible, for example as parameters for functions which won't add another reference. Another example is CEpgInfoTag which has a pointer back to its parent; this pointer does not affect lifetime, and using std::shared_ptr here would create a reference loop and never free any memory.
Maybe I missed a location where std::shared_ptr is necessary, and we can add these once we identify them, but in any case this commit fixes several bad crash bugs.

ksooo commented:
@MaxKellermann all fine. Just wanted to check that you thought about this as well. Much appreciated. 

This PR. is definitely an improvement. Stumbled over random crashes in that area as well in the past. :-/

+1 from my side.

ksooo commented:
Just a side-note: you seem to like to use "auto" even for trivial types like CEpgPtr. IMO this makes code less readable and thereby contradicts the purpose of auto. But ofc this might be just my very personal opinion.

MaxKellermann commented:
> you seem to like to use "auto" even for trivial types like CEpgPtr.

I like to use "auto" everywhere, but that's personal preference. If Kodi's coding style says I should not use it here, I'll change my patch accordingly. You decide.

ksooo commented:
Well, not exactly covered by our coding guidelines, so no hard rule for that. But I would be happy if you'd change this, as for the Kodi code area you touched we use auto only for non-trivial types.

MaxKellermann committed:
EpgContainer: use shared_ptr to manage CEpg pointers

This fixes a crash that can occur when one thread is inside
PersistAll() and the other is inside DeleteEpg().  The former does not
lock the mutex, as it operates on a std::map copy with CEpg pointers.
These CEpg instances however are deleted by the other thread.

Background information:

CEpgContainer::m_critSection protects the std::map m_epgs which
contains CEpg pointers.  All access to m_epgs must be protected with
m_critSection.  What the authors of that code didn't consider: if a
CEpg pointer escapes that protected code section, it can be
invalidated at any time.  Which implies: it is illegal to leak CEpg
pointers out of CEpgContainer::m_critSection protection!

That, however, is what commit 2ba1c43 implements, and what has been
done wrong many times before, for example GetById() and GetByChannel()
leak those pointers, explicitly.  All of that is wrong, and each one
of them can crash Kodi.

So what this commit does is ensure that all CEpg instances remain
valid as long as somebody is still holding a pointer, by wrapping all
CEpg pointers in std::shared_ptr.

MaxKellermann commented:
Changed a few "auto" to "CEpgPtr" and rebased on current master.

ksooo commented:
@MaxKellermann Thank's alot.

jenkins build this please

ksooo commented:
Jenkins errors are not related to this PR (bin-addon screensaver.matrixtrails on LINUX, checkout error on ATV2)
